test/test_topology.ml

```diff
@@ -0,0 +1,204 @@
+open Topology
+
+module D = Debug.Make (struct
+  let name = "test_topology"
+end)
+
+let make_numa ~numa ~sockets:_ ~cores =
+  let distances =
+    Array.init numa (fun i ->
+        Array.init numa (fun j ->
+            if i = j then 10 else 10 + (11 * abs (j - i))))
+  in
+  let cores_per_numa = cores / numa in
+  let cpu_to_node = Array.init cores (fun core -> core / cores_per_numa) in
+  NUMA.v ~distances ~cpu_to_node
+
+let make_numa_assymetric ~cores_per_numa =
+  (* e.g. AMD Opteron 6272 *)
+  let numa = 8 in
+  let distances =
+    [| [|10; 16; 16; 22; 16; 22; 16; 22|]
+     ; [|16; 10; 22; 16; 16; 22; 22; 17|]
+     ; [|16; 22; 10; 16; 16; 16; 16; 16|]
+     ; [|22; 16; 16; 10; 16; 16; 22; 22|]
+     ; [|16; 16; 16; 16; 10; 16; 16; 22|]
+     ; [|22; 22; 16; 16; 16; 10; 22; 16|]
+     ; [|16; 22; 16; 22; 16; 22; 10; 16|]
+     ; [|22; 16; 16; 22; 22; 16; 16; 10|] |]
+  in
+  let cpu_to_node =
+    Array.init (cores_per_numa * numa) (fun core -> core / cores_per_numa)
+  in
+  NUMA.v ~distances ~cpu_to_node
+
+type t = {worst: int; average: float; nodes: NUMA.node list; best: int}
+
+let pp =
+  Fmt.(
+    Dump.record
+      [ Dump.field "worst" (fun t -> t.worst) int
+      ; Dump.field "average" (fun t -> t.average) float
+      ; Dump.field "nodes" (fun t -> t.nodes) (Dump.list NUMA.pp_dump_node)
+      ; Dump.field "best" (fun t -> t.best) int ])
+
+let sum_costs l =
+  D.debug "====" ;
+  List.fold_left
+    (fun accum cost ->
+      { worst= max accum.worst cost.worst
+      ; average= accum.average +. cost.average
+      ; nodes= cost.nodes @ accum.nodes
+      ; best= min accum.best cost.best })
+    {worst= min_int; average= 0.; nodes= []; best= max_int}
+    l
+
+let vm_access_costs host all_vms (vcpus, nodes, cpuset) =
+  let nodes = List.of_seq nodes in
+  let all_vms = ((vcpus, nodes), cpuset) :: all_vms in
+  let n = List.length nodes in
+  let costs =
+    cpuset |> CPUSet.elements
+    |> List.map (fun c ->
+           let distances =
+             List.map
+               (fun node ->
+                 let d = NUMA.distance host (NUMA.node_of_cpu host c) node in
+                 let (NUMA.Node nodei) = node in
+                 D.debug "CPU %d <-> Node %d distance: %d" c nodei d ;
+                 d)
+               nodes
+           in
+           D.debug "Distances: %s"
+             (List.map string_of_int distances |> String.concat ",") ;
+           let worst = List.fold_left max 0 distances in
+           let best = List.fold_left min max_int distances in
+           let average = float (List.fold_left ( + ) 0 distances) /. float n in
+           {worst; best; nodes= []; average})
+    |> sum_costs
+  in
+  D.debug "Costs: %s" (Fmt.to_to_string pp costs) ;
+  let cpus = float @@ CPUSet.cardinal cpuset in
+  let nodes =
```

Comment:
have we enabled compiler warnings for shadowing?

my opinion is that we should

Reply:
Shadowing warnings only exist for `open` statements, I haven't found a warning for declaring a value with a name that already exists, I think it is done quite often for wrapping a recursive function and passing an initial value, etc.